### PR TITLE
WIP sync, final PR before integration

### PR DESCRIPTION
**Pull request by alonmuroch**
Including full sync implementation + unit tests

### CODE REVIEW

File: ibft/ibft_network.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import "github.com/bloxapp/ssv/network"
+
+func (i *ibftImpl) listenToNetworkMessages() {
+	msgChan := i.network.ReceivedMsgChan()
+	go func() {
+		for msg := range msgChan {
+			i.msgQueue.AddMessage(&network.Message{
+				Lambda:        msg.Message.Lambda,
+				SignedMessage: msg,
+				Type:          network.NetworkMsg_IBFTType,
```
Comment (nivBlox) [inline ibft/ibft_network.go:12]: GetTopic is depend on the msg validatorPublicKey so we need to add it to all msg's

File: ibft/instance.go [hunk 2]
```diff
@@ -134,9 +146,23 @@ func (i *Instance) Start(inputValue []byte) {
...
-	return i.State.Stage
+// Stop will trigger a stop for the entire instance
+func (i *Instance) Stop() {
+	i.stopLock.Lock()
+	defer i.stopLock.Unlock()
+
+	i.stop = true
+	i.stopRoundChangeTimer()
+	i.SetStage(proto.RoundState_Stopped)
+	i.Logger.Info("stopping iBFT instance...")
```
Comment (nivBlox) [inline ibft/instance.go:157]: Do we need to remove instance in ibft wrapper too?

File: network/p2p/p2p.go [hunk 3]
```diff
@@ -250,13 +251,13 @@ func (n *p2pNetwork) listen(sub *pubsub.Subscription) {
 	}(sub)
 }
 
-// GetTopic return topic by validator public key
-func (n *p2pNetwork) GetTopic(msg *proto.SignedMessage) (*pubsub.Topic, error) {
-	if msg.Message == nil || msg.Message.ValidatorPk == nil{
+// getTopic return topic by validator public key
+func (n *p2pNetwork) getTopic(validatorPK []byte) (*pubsub.Topic, error) {
```
Comment (nivBlox) [inline network/p2p/p2p.go:255]: Should be public, you sure its not failing?

File: network/p2p/p2p.go [hunk 4]
```diff
@@ -250,13 +251,13 @@ func (n *p2pNetwork) listen(sub *pubsub.Subscription) {
...
+	if validatorPK == nil {
 		return nil, errors.New("missing Message or ValidatorPk in signMsg")
```
Comment (nivBlox) [inline network/p2p/p2p.go]: need to change to - 
"validatorPk is nil"

File: validator/validator.go [hunk 5]
```diff
@@ -44,11 +44,13 @@ func New(ctx context.Context, logger *zap.Logger, validatorShare *collections.Va
 	msgQueue := msgqueue.New()
 	ibfts := make(map[beacon.Role]ibft.IBFT)
 	ibfts[beacon.RoleAttester] = ibft.New(
+		logger,
 		ibftStorage,
 		network,
 		msgQueue,
 		&proto.InstanceParams{
 			ConsensusParams: proto.DefaultConsensusParams(),
+			IbftCommittee:   validatorShare.Committee,
```
Comment (nivBlox) [inline validator/validator.go:53]: we sure there is no place that using that?

File: ibft/ibft_network.go (hunk 1 above)
Comment (alonmuroch) [reply ibft/ibft_network.go:12]: not sure what you mean:?

File: ibft/instance.go (hunk 2 above)
Comment (alonmuroch) [reply ibft/instance.go:157]: no, a stopped instance is also the current instance. 
I did find out I didn't set current instance so I've fixed it.

File: validator/validator.go (hunk 5 above)
Comment (alonmuroch) [reply validator/validator.go:53]: I've added an issue to refactor ibft committee #67

File: network/p2p/p2p.go (hunk 3 above)
Comment (alonmuroch) [reply network/p2p/p2p.go:255]: why should it be public? it's not part of the network interface.

File: ibft/ibft_network.go (hunk 1 above)
Comment (nivBlox) [reply ibft/ibft_network.go:12]: we need to pass validatorPublicKey for the GetTopic purpose

Comment (alonmuroch) [reply ibft/ibft_network.go:12]: you have that inside the signed msg

**Review by liorrutenberg — APPROVED** [review]

**Pull request merged**